File: content/tracks/dsa/problems/lc-0155-min-stack/solution.go
package main

// minStackEntry keeps a value and the minimum of the stack up to and including it.
type minStackEntry struct {
	val     int
	minimum int
}

type MinStack struct {
	entries []minStackEntry
}

func Constructor() MinStack {
	return MinStack{}
}

func (this *MinStack) Push(val int) {
	minimum := val
	if len(this.entries) > 0 && this.entries[len(this.entries)-1].minimum < val {
		minimum = this.entries[len(this.entries)-1].minimum
	}
	this.entries = append(this.entries, minStackEntry{val: val, minimum: minimum})
}

func (this *MinStack) Pop() {
	this.entries = this.entries[:len(this.entries)-1]
}

func (this *MinStack) Top() int {
	return this.entries[len(this.entries)-1].val
}

func (this *MinStack) GetMin() int {
	return this.entries[len(this.entries)-1].minimum
}

File: content/tracks/dsa/problems/lc-0155-min-stack/solution.py
from typing import List, Tuple


class MinStack:
    def __init__(self) -> None:
        # Each entry keeps its value and the minimum of the stack up to and including it.
        self.entries: List[Tuple[int, int]] = []

    def push(self, val: int) -> None:
        minimum = val if not self.entries else min(val, self.entries[-1][1])
        self.entries.append((val, minimum))

    def pop(self) -> None:
        self.entries.pop()

    def top(self) -> int:
        return self.entries[-1][0]

    def getMin(self) -> int:
        return self.entries[-1][1]
